src/main/java/de/atextor/turtle/formatter/TurtleFormatter.java

```diff
@@ -222,18 +244,63 @@ private State buildInitialState( final Model model, final OutputStream outputStr
      * Anonymous resources that are referred to more than once need to be given an internal id and
      * can not be serialized using [ ] notation.
      *
-     * @param model the input model
+     * @param model                    the input model
+     * @param currentState the state
      * @return the set of anonymous resources that are referred to more than once
      */
-    private Set<Resource> anonymousResourcesThatNeedAnId( final Model model ) {
-        return model.listObjects().toList().stream()
+    private Set<Resource> anonymousResourcesThatNeedAnId( final Model model, State currentState) {
+        Set<Resource> identifiedResources = new HashSet<>();
+        identifiedResources.addAll(currentState.identifiedAnonymousResources.keySet()); //needed for cycle detection
+        Set<Resource> candidates = model.listObjects().toList().stream()
             .filter( RDFNode::isResource )
             .map( RDFNode::asResource )
-            .filter( RDFNode::isAnon )
-            .filter( object -> statements( model, null, object ).size() > 1 )
-            .collect( Collectors.toSet() );
+            .filter( RDFNode::isAnon ).collect(Collectors.toSet());
+        candidates.removeAll(currentState.getBlankNodeMetadata().getLabeledBlankNodes());
+        List<Resource> candidatesInOrder = new ArrayList<>();
+        candidatesInOrder.addAll(
+                        new ArrayList<>(currentState.getBlankNodeMetadata().getLabeledBlankNodes())
+                        .stream()
+                        .sorted( currentState.getRDFNodeComparatorFactory().comparator())
+                        .toList());
+        candidatesInOrder.addAll(
+                        candidates
+                        .stream()
+                        .sorted( currentState.getRDFNodeComparatorFactory().comparator())
+                        .toList());
```

Comment:
Can this be simplified to the following?

```java
List<Resource> candidatesInOrder = Streams.concat(
     currentState.getBlankNodeMetadata().getLabeledBlankNodes().stream()
       .sorted( currentState.getRDFNodeComparatorFactory().comparator() ),
     candidates.stream()
       .sorted( currentState.getRDFNodeComparatorFactory().comparator() ) )
  .toList();
```

Reply:
done